library/core/src/char/methods.rs

```diff
@@ -331,14 +332,12 @@ impl char {
     #[inline]
     pub fn to_digit(self, radix: u32) -> Option<u32> {
         assert!(radix <= 36, "to_digit: radix is too high (maximum 36)");
-
+        const ASCII_DIGIT_MASK: u32 = 0b11_0000;
         // the code is split up here to improve execution speed for cases where
         // the `radix` is constant and 10 or smaller
-        let val = if radix <= 10 {
-            match self {
-                '0'..='9' => self as u32 - '0' as u32,
-                _ => return None,
-            }
+        let val = if likely(radix <= 10) {
+            // If not a digit, a number greater than radix will be created.
+            self as u32 ^ ASCII_DIGIT_MASK
```

Comment:
Maybe it's easier to follow when using `wrapping_sub` instead of xor. Then it also doesn't depend on the exact bit pattern of the ascii digits, but only on the fact that `'0'..='9'` are consecutive:
```suggestion
            (self as u32).wrapping_sub('0' as u32)
```

Reply:
That's cute and sligthly less mysterious.